Automattic/jetpack
# My Jetpack: Improve product cache handling on Jetpack AI

Fixes #34347.

## Proposed changes:

* Busts the product cache data when the tiers are enabled but the `price_tier_list` is empty (meaning that the site does not have tier prices yet, meaning it needs a fresh product information)
* If the new data does not contain tiers too (unexpected), just fails gracefully without returning any price, showing the interstitial without the price
* To allow it, include support for cache busting on the `Wpcom_Products::get_product()` function

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






### Fill the cache with production product data (no tiers)

* Launch a new JN site running this branch and connect it (or remove any Jetpack AI subscription from you local testing site)
* Go to `Jetpack > My Jetpack`, scroll to the Jetpack AI card and click the purchase button
* You should see the Jetpack AI product interstitial with the price for the current unlimited plan
* Now your cache is filled with production values (no tiers)

### Testing with the tiers

* Now go to your sandbox, apply D126642-code and rebase it with trunk
* Point the site to your sandbox (remember to enable HTTP access to it if it's a JN site)
* Leave the tiers _disabled_ for now and confirm the Jetpack AI interstitial is still showing the price for the legacy plan
* Enable the tiers using the filter:

```
add_filter( 'jetpack_ai_tier_plans_enabled', '__return_true' );
```

* Reload the Jetpack AI interstitial and confirm that you see the price for the 100 requests tier (the interstitial will state "100 requests" on the feature list)
   * If you have QueryMonitor installed on the test site, you can inspect the `HTTP API Calls` section and confirm a request was made to `https://public-api.wordpress.com/rest/v1.1/sites/BLOG_ID/products/`, to refresh the cache:

<img width="500" alt="Screenshot 2023-12-01 at 17 04 08" src="https://github.com/Automattic/jetpack/assets/6760046/b608fe89-e4ba-48fe-ae19-9f4b9c69cd1a">

* Buy the 100 requests tier (make sure you are sandboxing the store) and then go back to the product interstitial (`Jetpack > My Jetpack`, scroll the page to the Jetpack AI card, click `Purchase`)
* Confirm that now you see the price for the 200 requests tier (before this change, you would see the price for the legacy unlimited plan, a wrong behavior)
* Buy the 200 requests tier and proceed with the test again, to see if the 500 requests tier price is showed correctly too
* No need to test on all the tiers, but you can if you would like to
